app_dart/lib/src/request_handlers/task_status_subscription.dart

```diff
@@ -0,0 +1,127 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:cocoon_service/src/model/luci/push_message.dart';
+import 'package:cocoon_service/src/request_handling/subscription_handler.dart';
+import 'package:gcloud/db.dart';
+import 'package:meta/meta.dart';
+
+import '../model/appengine/commit.dart';
+import '../model/appengine/task.dart';
+import '../request_handling/authentication.dart';
+import '../request_handling/body.dart';
+import '../request_handling/exceptions.dart';
+import '../service/cache_service.dart';
+import '../service/config.dart';
+import '../service/datastore.dart';
+import '../service/logging.dart';
+
+/// An endpoint for listening to build updates for task updates.
+///
+/// The PubSub subscription is set up here:
+/// https://cloud.google.com/cloudpubsub/subscription/detail/luci-builds?project=flutter-dashboard&tab=overview
+///
+/// This endpoint is responsible for updating Datastore with the result of tasks from LUCI.
+@immutable
+class TaskStatusSubscription extends SubscriptionHandler {
+  const TaskStatusSubscription(
+    CacheService cache,
+    Config config,
+    AuthenticationProvider authProvider, {
+    @visibleForTesting this.datastoreProvider = DatastoreService.defaultProvider,
+  }) : super(cache: cache, config: config, authProvider: authProvider, topicName: 'luci-builds');
+
+  final DatastoreServiceProvider datastoreProvider;
+
+  @override
+  Future<Body> post() async {
+    final String data = message.data!;
+    final BuildPushMessage buildPushMessage =
```

Comment:
`BuildPushMessage` is specific to LUCI. Let's create a dedicated model for this as referencing `userData` requires manual validation

Reply:
Should we add some validation code to `BuildPushMessage`? How generic would we want the new model to be?

Reply:
`userData` is intended to be generic. For example, we change it based on presubmit/postsubmit, and if its original or a retry (which creates 4 possible versions).

Something like this should suffice

```dart
class ResultsUpdate extends JsonBody {
  final String sha;
  final String branch;
  final RepositorySlug slug;
  final String name;
  final String result;
  final DateTime? started;
  final DateTime? finished;
}
```